### DISCUSSION TITLE
Calculations.CalculateEhlersStochastic Not giving same Info as TraderView

### DISCUSSION THREAD

**Discussion by 200317314** [ooples/OoplesFinance.StockIndicators#53] [Q&A]
I am using the same inputs as Ooples on TraderView and using this library, but the library is giving me different numbers, and when I checked, it appears that the library is incorrect, has anyone else faced this?

I am using Ehlers2PoleSuperSmootherFilterV1 to smooth it

**Comment by ooples** [comment]
@200317314 Please share any code you have run on your end, link to the tradingview script you are comparing to as well as the results you are getting when using the library method. I publish a bunch of scripts on tradingview under the name cheatcountry so there shouldn't be any difference between my library and scripts on tradingview but mistakes could always happen of course

**Reply by 200317314** [reply]
@ooples so the EhlersStochastic is a bit off, this is the pine script from tradingview
`study("Ehlers Stochastic", shorttitle="EStoch")

highpassLength = input(title="High-Pass Filter Length", type=integer, defval=48)
stochLength = input(title="Stochastic Length", type=integer, defval=20)
ssfLength = input(title="Super Smoother Filter Length", type=integer, defval=10)
numberOfPoles = input(title="Super Smoother Filter Poles", type=integer, defval=2, options=[2, 3])
obLevel = input(title="Overbought Level", type=float, defval=0.8, step=0.1)
osLevel = input(title="Oversold Level", type=float, defval=0.2, step=0.1)
src = input(title="Source", type=source, defval=close)
highlightBreakouts = input(title="Highlight Overbought/Oversold Breakouts ?", type=bool, defval=true)

PI = 2 * asin(1)

_ssf2(src, length) =>
    arg = sqrt(2) * PI / length
    a1 = exp(-arg)
    b1 = 2 * a1 * cos(arg)
    c2 = b1
    c3 = -pow(a1, 2)
    c1 = 1 - c2 - c3
    
    ssf = 0.0
    ssf := c1 * src + c2 * nz(ssf[1]) + c3 * nz(ssf[2])

_ssf3(src, length) =>
    arg = PI / length
    a1 = exp(-arg)
    b1 = 2 * a1 * cos(1.738 * arg)
    c1 = pow(a1, 2)

    coef2 = b1 + c1
    coef3 = -(c1 + b1 * c1)
    coef4 = pow(c1, 2)
    coef1 = 1 - coef2 - coef3 - coef4

    ssf = 0.0
    ssf := coef1 * src + coef2 * nz(ssf[1]) + coef3 * nz(ssf[2]) + coef4 * nz(ssf[3])

_rf(src, highpassLength, ssfLength, numberOfPoles) =>
    alphaArg = 2 * PI / (highpassLength * sqrt(2))
    
    alpha = 0.0
    alpha := cos(alphaArg) != 0
         ? (cos(alphaArg) + sin(alphaArg) - 1) / cos(alphaArg)
         : nz(alpha[1])
    
    highpass = 0.0
    highpass := pow(1 - (alpha / 2), 2) * (src - 2 * nz(src[1]) + nz(src[2])) + 2 * (1 - alpha) * nz(highpass[1]) - pow(1 - alpha, 2) * nz(highpass[2])
    arg = (highpass + nz(highpass[1])) / 2

    numberOfPoles == 2
         ? _ssf2(arg, ssfLength)
         : _ssf3(arg, ssfLength)

_stoch(rf, stochLength, ssfLength, numberOfPoles) =>
    estoch = fixnan(stoch(rf, rf, rf, stochLength) / 100)
    estoch := na(estoch) ? 0 : estoch

    arg = (estoch + nz(estoch[1])) / 2
    
    numberOfPoles == 2
         ? _ssf2(arg, ssfLength)
         : _ssf3(arg, ssfLength)

erf = _rf(src, highpassLength, ssfLength, numberOfPoles)
estoch = _stoch(erf, stochLength, ssfLength, numberOfPoles)

estochColor = estoch > obLevel ? #0ebb23 : estoch < osLevel ? red : #f4b77d
plot(estoch, title="EStoch", linewidth=2, color=estochColor, transp=0)

transparent = color(white, 100)

maxLevelPlot = hline(1, title="Max Level", linestyle=dotted, color=transparent)
obLevelPlot = hline(obLevel, title="Overbought Level", linestyle=dotted)
hline(0.5, title="Middle Level", linestyle=dotted)
osLevelPlot = hline(osLevel, title="Oversold Level", linestyle=dotted)
minLevelPlot = hline(0, title="Min Level", linestyle=dotted, color=transparent)

fill(obLevelPlot, osLevelPlot, color=purple, transp=95)

obFillColor = estoch > obLevel and highlightBreakouts ? green : transparent
osFillColor = estoch < osLevel and highlightBreakouts ? red : transparent

fill(maxLevelPlot, obLevelPlot, color=obFillColor, transp=90)
fill(minLevelPlot, osLevelPlot, color=osFillColor, transp=90)
`

I am using settings 48, 3, 7 on the the 2 poles
Trading view im getting on QQQ -0.0077
When I use your function
`Calculations.CalculateEhlersStochastic(stockData, OoplesFinance.StockIndicators.Enums.MovingAvgType.Ehlers2PoleSuperSmootherFilterV1, 48, 3, 7);`
I get -0.0329 and pretty much ever value is off

**Reply by 200317314** [reply]
https://www.tradingview.com/script/iNw44glL-Ehlers-Stochastic/

**Reply by ooples** [reply]
@200317314 It is possible of course that I made a mistake but I double checked my source code to the link you sent (written by a friend of mine) and I'm not seeing any differences. I might have missed something so please let me know if you spot any differences in my version. I also checked Ehlers work and my version matches his exactly

**Reply by 200317314** [reply]
@ooples interesting, why am I getting the wrong data? I have the same data
![image](https://github.com/ooples/OoplesFinance.StockIndicators/assets/31968510/cccc8aa8-d3f3-4742-abc6-61028ee2d894)
Are you using close or HL? and which supersmoother should I use?

**Reply by 200317314** [reply]
@ooples I checked the code it is using close, but i cannot for the life of me figure out why its not giving me the same numbers

**Comment by ooples** [comment]
@200317314 My version uses close as well by default but you can also customize my version by changing the input source

**Reply by 200317314** [reply]
@ooples I played around with the source and it uses close, and I tried other smoothing and SMA/EMA I am not sure why my outputvalues are different, if you get a chance to use the traderview link, and use the 48,3,7 and try it out with the calculation you will see they are not the same.

**Comment by ooples** [comment]
I'm not doubting you at all. I'm just pointing out that comparing the
source code, it looks identical to me comparing my version with their
version. Please help me by taking a look yourself and seeing if I'm missing
anything. Just compare the two versions and see if you spot any differences
between how I do it and how the tradingview script does it

**Reply by 200317314** [reply]
@ooples 
 I found some issues while comparing tradingview with yours, not sure if I am correct or mistaken but some things ive seen are

in CalculateEhlersRoofingFilterV1() when we calc the CalculateEhlersHighPassFilterV1
TradingView
highpass := pow(1 - (alpha / 2), 2) * (src - 2 * nz(src[1]) + nz(src[2])) + 2 * (1 - alpha) * nz(highpass[1]) - pow(1 - alpha, 2) * nz(highpass[2]) arg = (highpass + nz(highpass[1])) / 2

Yours looks like this
ar highPass = (pow1 * (currentValue - (2 * prevValue1) + prevValue2)) + (2 * (1 - alpha) * prevHp1) - (pow2 * prevHp2); highPassList.AddRounded(highPass);

Another thing is on tradingview when we calc the _ssf2
`arg = sqrt(2) * PI / length
a1 = exp(-arg)
b1 = 2 * a1 * cos(arg)
c2 = b1
c3 = -pow(a1, 2)
c1 = 1 - c2 - c3

ssf = 0.0
ssf := c1 * src + c2 * nz(ssf[1]) + c3 * nz(ssf[2])`
It looks like you calc the supersmotherfiilter BEFORE you get the highpass instead of after like tradingview

After getting the erf, im not sure when you calculate the estoch = _stoch(erf, stochLength, ssfLength, numberOfPoles)?

**Reply by 200317314** [reply]
@ooples were you able to take a look at this by chance?

**Comment by 200317314** [comment]
@ooples  I know you are tired of hearing from me, but I found some real differences
in `CalculateEhlersSuperSmootherFilter` you see
`var filt = (coeff1 * ((currentValue + prevValue) / 2)) + (coeff2 * prevFilter1) + (coeff3 * prevFilter2);`

But in TraderView it looks like this
`ssf := c1 * src + c2 * nz(ssf[1]) + c3 * nz(ssf[2])`

shouldnt your code just be `currentValue` and not `(currentValue + prevValue) / 2` ?

you are also multiplying the prevFilters by the coefficients and im not sure why?

**Reply by 200317314** [reply]
@ooples further investigation shows that roofing filter you have works correctly, but the `var estoch2PoleList = GetMovingAverageList(stockData, maType, length2, arg2PoleList);` when you use `Ehlers2PoleSuperSmootherFilterV1` it does not work, it needs to be `EhlersSuperSmootherFilter` so the roofing filter needs the 2PoleV1 for some reason, but the smoothing on the stochastic needs normal, I still have not been able to get it to match up thought
